### PR TITLE
[OpenACC] Implement 'reduction' sema for compute constructs

### PR DESCRIPTION
**Pull request by erichkeane**
'reduction' has a few restrictions over normal 'var-list' clauses:

1- On parallel, a num_gangs can only have 1 argument when combined with reduction. These two aren't able to be combined on any other of the compute constructs however.

2- The vars all must be 'numerical data types' types of some sort, or a 'composite of numerical data types'.  A list of types is given in the standard as a minimum, so we choose 'isScalar', which covers all of these types and keeps types that are actually numeric. Other compilers don't seem to implement the 'composite of numerical data types', though we do.

3- Because of the above restrictions, member-of-composite is not allowed, so any access via a memberexpr is disallowed. Array-element and sub-arrays (aka array sections) are both permitted, so long as they meet the requirements of #2.

This patch implements all of these for compute constructs.

### CODE REVIEW

File: clang/lib/Sema/SemaOpenACC.cpp [hunk 1]
```diff
@@ -426,6 +438,24 @@ SemaOpenACC::ActOnClause(ArrayRef<const OpenACCClause *> ExistingClauses,
...
 
+    // OpenACC 3.3 Section 2.5.4:
+    // A reduction clause may not appear on a parallel construct with a
+    // num_gangs clause that has more than one argument.
+    if (Clause.getDirectiveKind() == OpenACCDirectiveKind::Parallel &&
+        Clause.getIntExprs().size() > 1) {
+      auto *Parallel =
+          llvm::find_if(ExistingClauses, [](const OpenACCClause *C) {
+            return C->getClauseKind() == OpenACCClauseKind::Reduction;
+          });
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenACC.cpp:449]: `llvm::find_if(ExitingClauses, IsaPred<OpenACCReductionClause>)`?

File: clang/lib/Sema/SemaOpenACC.cpp [hunk 2]
```diff
@@ -706,6 +736,48 @@ SemaOpenACC::ActOnClause(ArrayRef<const OpenACCClause *> ExistingClauses,
...
+      break;
+
+    // OpenACC 3.3 Section 2.5.4:
+    // A reduction clause may not appear on a parallel construct with a
+    // num_gangs clause that has more than one argument.
+    if (Clause.getDirectiveKind() == OpenACCDirectiveKind::Parallel) {
+      auto NumGangsClauses =
+          llvm::make_filter_range(ExistingClauses, [](const OpenACCClause *C) {
+            return C->getClauseKind() == OpenACCClauseKind::NumGangs;
+          });
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenACC.cpp:753]: `llvm::make_filter_range(ExistingClauses, IsaPred<OpenACCNumGangsClause>)`?

File: clang/lib/Sema/SemaOpenACC.cpp [hunk 3]
```diff
@@ -715,6 +787,65 @@ SemaOpenACC::ActOnClause(ArrayRef<const OpenACCClause *> ExistingClauses,
...
+          << /*not class or struct*/ 0 << VarExpr->getType();
+      return ExprError();
+    }
+
+    if (!RD->isCompleteDefinition()) {
+      Diag(VarExpr->getExprLoc(), diag::err_acc_reduction_composite_type)
+          << /*incomplete*/ 1 << VarExpr->getType();
+      return ExprError();
+    }
+    if (isa<CXXRecordDecl>(RD) && !cast<CXXRecordDecl>(RD)->isAggregate()) {
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenACC.cpp:826]: `if (auto *CXXRD = dyn_cast<CXXRecordDecl>(RD); CXXRD && !CXXRD->`?

**Review by alexey-bataev — COMMENTED** [review]
ast printing tests?

File: clang/lib/Sema/SemaOpenACC.cpp (hunk 1 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaOpenACC.cpp:449]: That is awesome, thank you for showing me this!

File: clang/lib/Sema/SemaOpenACC.cpp (hunk 3 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaOpenACC.cpp:826]: Yep, better.

File: clang/test/AST/ast-print-openacc-compute-construct.cpp [hunk 4]
```diff
@@ -130,5 +130,33 @@ void foo() {
 //CHECK: #pragma acc parallel device_type(SomeStructImpl)
 #pragma acc parallel device_type (SomeStructImpl)
   while(true);
+
+//CHECK: #pragma acc parallel reduction(+: iPtr)
+#pragma acc parallel reduction(+: iPtr)
```
Comment (erichkeane) [inline clang/test/AST/ast-print-openacc-compute-construct.cpp:135]: AST print tests are here

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**